--- a/lib/dpkg/file.c
+++ b/lib/dpkg/file.c
@@ -182,6 +182,9 @@ file_slurp_fd(int fd, const char *filename, struct varbuf *vb,
               struct dpkg_error *err)
 {
 	struct stat st;
+#ifdef __wasi__
+	ssize_t bytes_read;
+#endif
 
 	if (fstat(fd, &st) < 0)
 		return dpkg_put_errno(err, _("cannot stat %s"), filename);
@@ -196,9 +199,16 @@ file_slurp_fd(int fd, const char *filename, struct varbuf *vb,
 		return 0;
 
 	varbuf_init(vb, st.st_size + 1);
+#ifdef __wasi__
+	bytes_read = fd_read(fd, vb->buf, st.st_size);
+	if (bytes_read < 0)
+		return dpkg_put_errno(err, _("cannot read %s"), filename);
+	varbuf_trunc(vb, bytes_read);
+#else
 	if (fd_read(fd, vb->buf, st.st_size) < 0)
 		return dpkg_put_errno(err, _("cannot read %s"), filename);
 	varbuf_trunc(vb, st.st_size);
+#endif
 
 	return 0;
 }
